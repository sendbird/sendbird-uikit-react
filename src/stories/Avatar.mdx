import { Meta, Canvas, Controls, ArgsTable } from '@storybook/addon-docs';
import Avatar from '../ui/Avatar';
import * as AvatarStories from './Avatar.stories';

<Meta title="UI/Avatar" component={Avatar} of={AvatarStories} />

# Avatar

The `Avatar` component displays images or initials for user profiles. It supports custom dimensions and can display default avatars when no image source is provided. Utilize the Avatar component to enhance user identity representation in your application, providing a more personalized user experience.

<Canvas of={AvatarStories.DefaultAvatar} />
<Controls />

### Importing Avatar

To integrate the Avatar component into your project, use the following import statement:

```tsx
import Avatar from '@sendbird/uikit-react/ui/Avatar';
```
