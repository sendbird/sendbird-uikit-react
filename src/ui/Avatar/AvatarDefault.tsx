import React from 'react';
import pxToNumber from '../../utils/pxToNumber';
import Icon, { IconColors, IconTypes } from '../Icon';
import { classnames } from '../../utils/utils';

interface AvatarDefaultProps {
  width: string | number,
  height: string | number,
  text?: string,
}

const AvatarDefault = (props: AvatarDefaultProps) => {
  const { width, height, text } = props;

  let iconWidth = pxToNumber(width);
  let iconHeight = pxToNumber(height);
  if (typeof iconWidth === 'number' && !Number.isNaN(iconWidth)) {
    iconWidth *= 0.575;
  }
  if (typeof iconHeight === 'number' && !Number.isNaN(iconHeight)) {
    iconHeight *= 0.575;
  }

  return (
    <div
      className={classnames('sendbird-avatar-img--default', text && 'text')}
      style={{ width, height }}
    >
      {
        text
          ? <div
            className='sendbird-avatar-text'
          >{ text }</div>
          : <Icon
            type={IconTypes.USER}
            fillColor={IconColors.CONTENT}
            width={iconWidth}
            height={iconHeight}
          />
      }
    </div>
  );
};

export default AvatarDefault;
