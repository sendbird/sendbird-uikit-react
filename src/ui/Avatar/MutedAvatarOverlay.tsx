import './muted-avatar-overlay.scss';

import React, { ReactElement } from 'react';
import Icon, { IconTypes, IconColors } from '../Icon';

interface Props {
  height?: number;
  width?: number;
}

export default function MutedAvatarOverlay(props: Props): ReactElement {
  const { height = 24, width = 24 } = props;
  return (
    <div
      className="sendbird-muted-avatar"
      style={{
        height: `${height}px`,
        width: `${width}px`,
      }}
    >
      <div className="sendbird-muted-avatar__icon">
        <div
          className="sendbird-muted-avatar__bg"
          style={{
            height: `${height}px`,
            width: `${width}px`,
          }}
        />
        <Icon
          type={IconTypes.MUTE}
          fillColor={IconColors.WHITE}
          width={`${height - 8}px`}
          height={`${width - 8}px`}
        />
      </div>
    </div>
  );
}
