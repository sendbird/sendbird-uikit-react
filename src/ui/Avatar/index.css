.sendbird-avatar {
  background-color: red;
}
