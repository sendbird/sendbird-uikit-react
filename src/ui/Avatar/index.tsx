import React, { ReactElement, RefObject } from 'react';
import ImageRenderer from '../ImageRenderer';
import './index.scss';
import { uuidv4 } from '../../utils/uuid';
import AvatarDefault from './AvatarDefault';

const imageRendererClassName = 'sendbird-avatar-img';

interface AvatarInnerProps {
  height: string | number,
  width: string | number,
  src?: string | Array<string>,
  alt?: string,
  customDefaultComponent?({ width, height }: { width: number | string, height: number | string }): ReactElement;
}

export const AvatarInner = ({
  src = '',
  alt = '',
  height,
  width,
  customDefaultComponent,
}: AvatarInnerProps): ReactElement => {
  const defaultComponent = () => customDefaultComponent
    ? customDefaultComponent({ width, height })
    : <AvatarDefault width={width} height={height} />;

  if (typeof src === 'string') {
    return (
      <ImageRenderer
        className={imageRendererClassName}
        url={src}
        height={height}
        width={width}
        alt={alt}
        defaultComponent={defaultComponent}
      />
    );
  }

  if (src && src.length) {
    if (src.length === 1) {
      return (
        <ImageRenderer
          className={imageRendererClassName}
          url={src[0]}
          height={height}
          width={width}
          alt={alt}
          defaultComponent={defaultComponent}
        />
      );
    }

    if (src.length === 2) {
      return (
        <div className="sendbird-avatar--inner__two-child">
          <ImageRenderer
            className={imageRendererClassName}
            url={src[0]}
            height={height}
            width={width}
            alt={alt}
            defaultComponent={defaultComponent}
          />
          <ImageRenderer
            className={imageRendererClassName}
            url={src[1]}
            height={height}
            width={width}
            alt={alt}
            defaultComponent={defaultComponent}
          />
        </div>
      );
    }

    if (src.length === 3) {
      return (
        <>
          <div className="sendbird-avatar--inner__three-child--upper">
            <ImageRenderer
              className={imageRendererClassName}
              url={src[0]}
              height={height}
              width={width}
              alt={alt}
              defaultComponent={defaultComponent}
            />
          </div>
          <div className="sendbird-avatar--inner__three-child--lower">
            <ImageRenderer
              className={imageRendererClassName}
              url={src[1]}
              height={height}
              width={width}
              alt={alt}
              defaultComponent={defaultComponent}
            />
            <ImageRenderer
              className={imageRendererClassName}
              url={src[2]}
              height={height}
              width={width}
              alt={alt}
              defaultComponent={defaultComponent}
            />
          </div>
        </>
      );
    }

    return (
      <div className="sendbird-avatar--inner__four-child">
        {
          src.slice(0, 4)
            .map((i) => (
              <ImageRenderer
                className={imageRendererClassName}
                url={i}
                height={height}
                width={width}
                alt={alt}
                key={uuidv4()}
                defaultComponent={defaultComponent}
              />
            ))
        }
      </div>
    );
  }
  // default img
  return (
    <ImageRenderer
      className={imageRendererClassName}
      url=""
      height={height}
      width={width}
      alt={alt}
      defaultComponent={defaultComponent}
    />
  );
};

interface AvatarProps {
  className?: string | Array<string>,
  height?: string | number,
  width?: string | number,
  zIndex?: string | number,
  left?: string,
  src?: string | Array<string>,
  alt?: string,
  onClick?(): void,
  customDefaultComponent?({ width, height }: { width: number | string, height: number | string }): ReactElement;
}

function Avatar(
  {
    className = '',
    src = '',
    alt = '',
    width = '56px',
    height = '56px',
    zIndex = 0,
    left = '0px',
    onClick,
    customDefaultComponent,
  }: AvatarProps,
  ref?: RefObject<HTMLDivElement>,
): ReactElement {
  return (
    <div
      className={[
        ...(Array.isArray(className) ? className : [className]),
        'sendbird-avatar',
      ].join(' ')}
      role="button"
      ref={ref}
      style={{
        height,
        width,
        zIndex,
        left,
      }}
      onClick={onClick}
      onKeyDown={onClick}
      tabIndex={0}
    >
      <AvatarInner
        src={src}
        width={width}
        height={height}
        alt={alt}
        customDefaultComponent={customDefaultComponent}
      />
    </div>
  );
}

export default React.forwardRef(Avatar);
