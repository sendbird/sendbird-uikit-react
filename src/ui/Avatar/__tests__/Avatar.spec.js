import React from 'react';
import { render } from '@testing-library/react';

import Avatar from '../index';

const src = 'https://avatars3.githubusercontent.com/u/46333979?s=460&v=4';
const src1 = 'https://avatars1.githubusercontent.com/u/1384313?s=460&v=4';
const src2 = 'https://avatars2.githubusercontent.com/u/11382805?s=460&v=4';
const src3 = '';

describe('ui/Avatar', () => {
  it('should render default image if src is empty', function() {
    const { asFragment } = render(
      <Avatar />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should do a snapshot test of the Avatar DOM with single element', function() {
    const { asFragment } = render(
      <Avatar src={src} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should do a snapshot test of the Avatar DOM with two elements', function() {
    const { asFragment } = render(
      <Avatar src={[src, src1]} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should do a snapshot test of the Avatar DOM with three elements', function() {
    const { asFragment } = render(
      <Avatar src={[src, src1, src2]} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should do a snapshot test of the Avatar DOM with four elements', function() {
    const { asFragment } = render(
      <Avatar src={[src, src1, src2, src3]} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
