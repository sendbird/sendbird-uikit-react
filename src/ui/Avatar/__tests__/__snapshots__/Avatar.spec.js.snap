// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/Avatar should do a snapshot test of the Avatar DOM with four elements 1`] = `
<DocumentFragment>
  <div
    class=" sendbird-avatar"
    role="button"
    style="height: 56px; width: 56px; z-index: 0; left: 0px;"
    tabindex="0"
  >
    <div
      class="sendbird-avatar--inner__four-child"
    >
      <div
        class="sendbird-avatar-img sendbird-image-renderer"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
      >
        <div
          class="sendbird-image-renderer__image"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://avatars3.githubusercontent.com/u/46333979?s=460&v=4);"
        />
        <img
          alt=""
          class="sendbird-image-renderer__hidden-image-loader"
          src="https://avatars3.githubusercontent.com/u/46333979?s=460&v=4"
        />
      </div>
      <div
        class="sendbird-avatar-img sendbird-image-renderer"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
      >
        <div
          class="sendbird-image-renderer__image"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://avatars1.githubusercontent.com/u/1384313?s=460&v=4);"
        />
        <img
          alt=""
          class="sendbird-image-renderer__hidden-image-loader"
          src="https://avatars1.githubusercontent.com/u/1384313?s=460&v=4"
        />
      </div>
      <div
        class="sendbird-avatar-img sendbird-image-renderer"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
      >
        <div
          class="sendbird-image-renderer__image"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://avatars2.githubusercontent.com/u/11382805?s=460&v=4);"
        />
        <img
          alt=""
          class="sendbird-image-renderer__hidden-image-loader"
          src="https://avatars2.githubusercontent.com/u/11382805?s=460&v=4"
        />
      </div>
      <div
        class="sendbird-avatar-img sendbird-image-renderer"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
      >
        <div
          class="sendbird-image-renderer__image"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url();"
        />
        <img
          alt=""
          class="sendbird-image-renderer__hidden-image-loader"
          src=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ui/Avatar should do a snapshot test of the Avatar DOM with single element 1`] = `
<DocumentFragment>
  <div
    class=" sendbird-avatar"
    role="button"
    style="height: 56px; width: 56px; z-index: 0; left: 0px;"
    tabindex="0"
  >
    <div
      class="sendbird-avatar-img sendbird-image-renderer"
      style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
    >
      <div
        class="sendbird-image-renderer__image"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://avatars3.githubusercontent.com/u/46333979?s=460&v=4);"
      />
      <img
        alt=""
        class="sendbird-image-renderer__hidden-image-loader"
        src="https://avatars3.githubusercontent.com/u/46333979?s=460&v=4"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`ui/Avatar should do a snapshot test of the Avatar DOM with three elements 1`] = `
<DocumentFragment>
  <div
    class=" sendbird-avatar"
    role="button"
    style="height: 56px; width: 56px; z-index: 0; left: 0px;"
    tabindex="0"
  >
    <div
      class="sendbird-avatar--inner__three-child--upper"
    >
      <div
        class="sendbird-avatar-img sendbird-image-renderer"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
      >
        <div
          class="sendbird-image-renderer__image"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://avatars3.githubusercontent.com/u/46333979?s=460&v=4);"
        />
        <img
          alt=""
          class="sendbird-image-renderer__hidden-image-loader"
          src="https://avatars3.githubusercontent.com/u/46333979?s=460&v=4"
        />
      </div>
    </div>
    <div
      class="sendbird-avatar--inner__three-child--lower"
    >
      <div
        class="sendbird-avatar-img sendbird-image-renderer"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
      >
        <div
          class="sendbird-image-renderer__image"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://avatars1.githubusercontent.com/u/1384313?s=460&v=4);"
        />
        <img
          alt=""
          class="sendbird-image-renderer__hidden-image-loader"
          src="https://avatars1.githubusercontent.com/u/1384313?s=460&v=4"
        />
      </div>
      <div
        class="sendbird-avatar-img sendbird-image-renderer"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
      >
        <div
          class="sendbird-image-renderer__image"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://avatars2.githubusercontent.com/u/11382805?s=460&v=4);"
        />
        <img
          alt=""
          class="sendbird-image-renderer__hidden-image-loader"
          src="https://avatars2.githubusercontent.com/u/11382805?s=460&v=4"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ui/Avatar should do a snapshot test of the Avatar DOM with two elements 1`] = `
<DocumentFragment>
  <div
    class=" sendbird-avatar"
    role="button"
    style="height: 56px; width: 56px; z-index: 0; left: 0px;"
    tabindex="0"
  >
    <div
      class="sendbird-avatar--inner__two-child"
    >
      <div
        class="sendbird-avatar-img sendbird-image-renderer"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
      >
        <div
          class="sendbird-image-renderer__image"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://avatars3.githubusercontent.com/u/46333979?s=460&v=4);"
        />
        <img
          alt=""
          class="sendbird-image-renderer__hidden-image-loader"
          src="https://avatars3.githubusercontent.com/u/46333979?s=460&v=4"
        />
      </div>
      <div
        class="sendbird-avatar-img sendbird-image-renderer"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
      >
        <div
          class="sendbird-image-renderer__image"
          style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url(https://avatars1.githubusercontent.com/u/1384313?s=460&v=4);"
        />
        <img
          alt=""
          class="sendbird-image-renderer__hidden-image-loader"
          src="https://avatars1.githubusercontent.com/u/1384313?s=460&v=4"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ui/Avatar should render default image if src is empty 1`] = `
<DocumentFragment>
  <div
    class=" sendbird-avatar"
    role="button"
    style="height: 56px; width: 56px; z-index: 0; left: 0px;"
    tabindex="0"
  >
    <div
      class="sendbird-avatar-img sendbird-image-renderer"
      style="width: 100%; min-width: min(400px, 56px); max-width: 400px;"
    >
      <div
        class="sendbird-image-renderer__image"
        style="width: 100%; min-width: min(400px, 56px); max-width: 400px; position: absolute; background-repeat: no-repeat; background-position: center; background-size: cover; background-image: url();"
      />
      <img
        alt=""
        class="sendbird-image-renderer__hidden-image-loader"
        src=""
      />
    </div>
  </div>
</DocumentFragment>
`;
